### PR TITLE
[US2167] Fixing Bug reported by coverity for iSCSI controller code

### PR DESCRIPTION
**Pull request by mynktl**
Changes : 
Fixing issues reported by the Coverity Scan.
List of issues : 

TA1366 | Memory leak in accept_mgmt_conns()
-- | --
TA1371 | In istgt_acceptor(), we unlock mutex without locking it
TA1375 | Mutex is used before it was initialized in istgt_init()
TA1378 | config_new need to be freed in error case in function istgt_reload()
TA1379 | host and port need to be freed in istgt_pg_match_all()
TA1409 | logfile need to be freed in istgt_connection_status() if it hit error.
TA1407 | Release mutex before sleep in istgt_iscsi_drop_old_conns()
TA1406 | Memory leak in istgt_iscsi_op_text()
TA1405 | Memory leak, dereferencing null pointer in istgt_chap_get_authinfo()
TA1400 | Dereferencing un-initialized pointer in istgt_iscsi_transfer_in_internal()
TA1399 | Dereferencing memroy after freeing it in istgt_remove_conn()
TA1392 | Release lock ""g_conns_mutex""  before call to sleep in istgt_iscsi_drop_all_conns()
TA1397 | Memory leak in istgt_iscsi_negotiate_params()
TA1398 | Return without releasing mutext in istgt_iscsi_check_values()
TA1416 | In function istgt_uctl_cmd_set(), dereferencing ""spec"" which might be NULL.
TA1424 | bp may overun memory in istgt_uctl_cmd_dump()
TA1425 | Memory leak in istgt_lu_disk_scsi_persistent_reserve_out()
TA1426 | Double locking of mutex in function istgt_lu_disk_scsi_report_target_port_groups()
TA1433 | Memory leak in istgt_lu_disk_lbwrite_ats()
TA1434 | Problem because of different return value and type of variable in istgt_lu_disk_lbwrite_same()
TA1439 | Memory leak in istgt_lu_disk_init()
TA1440 | Memory leak in istgt_get_xcopy_target
TA1441 | Memory leak in istgt_lu_disk_receive_copy_results()
TA1445 | Memory lean in istgt_lu_disk_lbxcopy

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

### CODE REVIEW

File: src/istgt.c [hunk 1]
```diff
@@ -329,21 +329,28 @@ istgt_pg_match_all(PORTAL_GROUP *pgp, CF_SECTION *sp)
...
-			return 0;
+			goto no_match;
 	}
 	label = istgt_get_nmval(sp, "Portal", i, 0);
 	portal = istgt_get_nmval(sp, "Portal", i, 1);
 	if (label != NULL || portal != NULL)
 		return 0;
 	return 1;
+
+no_match:
```
Comment (satbirchhikara) [inline src/istgt.c]: port and host may be leaked if we return at line 345 and 346.

**Review by satbirchhikara — APPROVED** [review]

File: src/istgt.c (hunk 1 above)
Comment (mynktl) [reply src/istgt.c]: ok

File: src/istgt.c [hunk 2]
```diff
@@ -317,33 +317,42 @@ istgt_add_portal_group(ISTGT_Ptr istgt, CF_SECTION *sp, int *pgp_idx)
...
 	label = istgt_get_nmval(sp, "Portal", i, 0);
 	portal = istgt_get_nmval(sp, "Portal", i, 1);
 	if (label != NULL || portal != NULL)
-		return 0;
-	return 1;
+		goto no_match;
+	ret = 1;
+
+no_match:
+	if (port)
```
Comment (vishnuitta) [inline src/istgt.c:351]: 'free'ing here doesn't look correct, as, this memory area may be still being pointed by data structures in 'istgt->config->section'

File: src/istgt.c [hunk 3]
```diff
@@ -2334,7 +2345,6 @@ istgt_acceptor(ISTGT_Ptr istgt)
 	event.events = EPOLLIN;
 	rc = epoll_ctl(epfd, EPOLL_CTL_ADD, istgt->sig_pipe[0], &event);
 	if (rc == -1) {
-		MTX_UNLOCK(&istgt->mutex);
```
Comment (vishnuitta) [inline src/istgt.c]: instead of removing UNLOCK code here, extend the lock i.e., remove UNLOCK at line#2315 of new file, and extend the lock to below this code.

File: src/istgt_iscsi.c [hunk 4]
```diff
@@ -6938,6 +6955,10 @@ istgt_remove_conn(CONN_Ptr conn)
 	}
 	sessConns = sess->connections;
 	MTX_UNLOCK(&sess->mutex);
+
+	lun = (sess->lu == NULL)? 0 : sess->lu->num;
```
Comment (vishnuitta) [inline src/istgt_iscsi.c]: 'lu_num' instead of 'lun'?

File: src/istgt_lu.c [hunk 5]
```diff
@@ -4136,7 +4137,7 @@ maintenance_io_worker(void *arg)
 	ISTGT_LU_DISK *spec = NULL;
 	ISTGT_LU_TASK_Ptr lu_task;
 	int qcnt;
-	int lun;
+	volatile int lun;
```
Comment (vishnuitta) [inline src/istgt_lu.c]: why this?

File: src/istgt_lu.c [hunk 6]
```diff
@@ -4179,6 +4180,8 @@ maintenance_io_worker(void *arg)
 	qcnt = 0;
 	pthread_cleanup_push(luw_cleanup, (void *)spec);
 
+	retry_runningstate_count = 0;
```
Comment (vishnuitta) [inline src/istgt_lu.c]: its already 0

File: src/istgt_lu.c [hunk 7]
```diff
@@ -4244,13 +4247,14 @@ maintenance_io_worker(void *arg)
 				}
 				MTX_UNLOCK(&spec->state_mutex);
 
+				rc = 0;
 				if (do_close == 1)
 					rc = istgt_lu_disk_close(spec->lu, spec->lun);
-				if (do_open == 1)
+				else if (do_open == 1)
```
Comment (vishnuitta) [inline src/istgt_lu.c]: not with 'else'.. it should be 'if'.. both can be set to 1

File: src/istgt_lu_ctl.c [hunk 8]
```diff
@@ -2492,16 +2483,19 @@ istgt_uctl_cmd_dump(UCTL_Ptr uctl)
 
 		istgt_lock_gconns();
 		MTX_LOCK(&lu->mutex);
-		for (j = 0; j < lu->maxmap; j++) {
+		/* limit host string to 2048 characters */
+		for (j = 0; j < lu->maxmap && rem > 256; j++) {
```
Comment (vishnuitta) [inline src/istgt_lu_ctl.c]: check for 'rem' > 256 should not be there. For below 'for' loop also.
In snprintf, change '256' to 'rem' in 2499 of old file

File: src/istgt_lu_ctl.c [hunk 9]
```diff
@@ -2492,16 +2483,19 @@ istgt_uctl_cmd_dump(UCTL_Ptr uctl)
...
-				for( x = 0; x < pgp->nportals; x++) {
+				for( x = 0; x < pgp->nportals && rem > 256; x++) {
 					ln = snprintf(bp, 256, " IP%d:%s  ", x+1, pgp->portals[x]->host);
-					if(ln <0)
+					if(ln < 0)
 						ln = 0;
-					else if(ln > rem)
-						ln = rem;
-					rem -= ln;
+					else if(ln > 0) {
```
Comment (vishnuitta) [inline src/istgt_lu_ctl.c]: these changes on 'ln' is not required

File: src/istgt_lu_disk.c [hunk 10]
```diff
@@ -1151,7 +1135,11 @@ istgt_lu_disk_init(ISTGT_Ptr istgt __attribute__((__unused__)), ISTGT_LU_Ptr lu)
 			    && spec->blocklen != 524288) {
 				ISTGT_ERRLOG("LU%d: LUN%d: invalid blocklen %"PRIu64"\n",
 				    lu->num, i, spec->blocklen);
-			error_return:
+error_return:
+#ifdef	REPLICATION
+				destroy_volume(spec);
+#endif
+spec_error:
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: by the time here, not all the variables are initialized

File: src/istgt_lu_disk.c [hunk 11]
```diff
@@ -1323,7 +1322,6 @@ istgt_lu_disk_init(ISTGT_Ptr istgt __attribute__((__unused__)), ISTGT_LU_Ptr lu)
 			}
 		} else {
 			ISTGT_ERRLOG("LU%d: LUN%d: unsupported format\n", lu->num, i);
-			goto error_return;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: its better if we return error here.. right?

File: src/istgt_lu_disk.c [hunk 12]
```diff
@@ -4171,11 +4178,7 @@ istgt_lu_disk_scsi_persistent_reserve_out(ISTGT_LU_DISK *spec,
 		conn->id, prout_sa[(sa > -1 && sa < PROUTSA_UNK) ? sa : PROUTSA_UNK], sa,
 		rkey, sarkey, spec_i_pt, all_tg_pt, aptpl, conn->initiator_port, spec->rsv_key);
 
-	rc = istgt_lu_disk_copy_reservation(spec_bkp, spec);
-	if (rc < 0) {
-		/* Ignore error */
-		bkp_success = -1;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c:4177]: this is in use later in the function. So, this is required.

File: src/istgt_lu_disk.c [hunk 13]
```diff
@@ -5113,6 +5116,13 @@ istgt_lu_disk_scsi_persistent_reserve_out(ISTGT_LU_DISK *spec,
 			xfree(spec_bkp->rsv_port);
 		xfree(spec_bkp);
 	}
+	if (ret == -1 && initiator_ports) {
+		for (i = 0; i < maxports; i++) {
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c:5126]: lets initialize maxports as 0

File: src/istgt_lu_disk.c [hunk 14]
```diff
@@ -115,7 +116,8 @@ extern clockid_t clockid;
 			markedForReturn = 1;	\
 			goto macroname;	\
 		}	\
-		return 0;	\
+		markedForReturn = 2;	\
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c:119]: enterblockingcall is called from many places. It doesn't seem that this error is handled everywhere.

Comment (vishnuitta) [reply src/istgt_lu_disk.c:119]: In above, can you change "IS_SPEC_READY" to "IS_SPEC_BUSY"? (not related to your PR)
current macro name is very confusing.

File: src/istgt_lu_disk.c [hunk 15]
```diff
@@ -8538,6 +8575,8 @@ istgt_lu_disk_cmd_queue_count(ISTGT_LU_Ptr lu, int *lun)
 				}\
 			}\
 			MTX_UNLOCK(&spec->lu_tmf_mutex[worker_id]);\
+			aborted = 0;				\
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: this initializations are not correct

File: src/istgt_lu_disk.c [hunk 16]
```diff
@@ -8680,6 +8719,8 @@ istgt_lu_disk_queue_start(ISTGT_LU_Ptr lu, int lun, int worker_id)
 		MTX_UNLOCK(&conn->result_queue_mutex);
 	}
 
+	msg = NULL;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: these initializations also not correct

File: src/istgt_lu_disk.c [hunk 17]
```diff
@@ -8973,6 +9011,7 @@ istgt_lu_disk_queue_start(ISTGT_LU_Ptr lu, int lun, int worker_id)
 error_return:
 	INFLIGHT_IO_CLEANUP;	
 error_return_no_cleanup:
+	printmsg = 0;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: not correct

File: src/istgt_misc.c [hunk 18]
```diff
@@ -243,8 +243,9 @@ poolfini(void)
 		mpools[i].endmem = NULL;
 
 	}
-	//if (gstart != NULL)
-	//	free(gstart);
+
+	if (gstart != NULL)
```
Comment (vishnuitta) [inline src/istgt_misc.c]: lets not do this as memory pointed by gstart is shared among pools, even though, pools related pointers are set to NULL in poolfini.
Leak here should be ok as process is going to die now.

File: src/replication.c [hunk 19]
```diff
@@ -757,14 +757,25 @@ create_replica_entry(spec_t *spec, int epfd, int mgmt_fd)
 	if (rc != 0) {
 		REPLICA_ERRLOG("pthread_mutex_init() failed err(%d) for "
 		    "replica(%s:%d)\n", rc, replica->ip, replica->port);
-		return NULL;
+		goto error;
```
Comment (vishnuitta) [inline src/replication.c:819]: pthread_mutex_init failed, so, should not be doing pthread_mutex_destroy at 'error' label

File: src/replication.c [hunk 20]
```diff
@@ -888,6 +899,7 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd)
 		REPLICA_ERRLOG("pthread_create(r_thread) failed for "
 		    "replica(%lu)\n", replica->zvol_guid);
 replica_error:
+		destroy_mempool(&replica->cmdq);
```
Comment (vishnuitta) [inline src/replication.c:972]: at various places, there is 'goto' to label 'replica_error' even before cmdq is initialized.

File: src/replication.c [hunk 21]
```diff
@@ -1606,32 +1618,47 @@ accept_mgmt_conns(int epfd, int sfd)
...
 		replica->replica_mgmt_dport = atoi(sbuf);
 #endif
+
 		rc = epoll_ctl(epfd, EPOLL_CTL_ADD, mgmt_fd, &event);
 		if(rc == -1) {
 			REPLICA_ERRLOG("epoll_ctl() failed on fd(%d), "
 			    "err(%d).. closing it.. for replica(%s:%d)\n",
 			    mgmt_fd, errno, replica->ip, replica->port);
 cleanup:
+			if (mevent1) {
```
Comment (vishnuitta) [inline src/replication.c]: lets delete mevent1 and mevent2 after doing EPOLL_CTL_DEL as the pointers are added to epoll.

**Review by vishnuitta — COMMENTED** [review]
General comment: As most of the code changes are in legacy and are seems to be related to one time or config change time, it would be less risky to avoid/ignore those changes.

File: src/istgt.c (hunk 2 above)
Comment (mynktl) [reply src/istgt.c:351]: In `istgt_parse_portal`, we are allocating memory for `port` and copy port number from istgt->config->section for only comparison purpose. So, we can free `port` and `host` here.

File: src/istgt.c (hunk 3 above)
Comment (mynktl) [reply src/istgt.c]: ok

File: src/istgt_iscsi.c (hunk 4 above)
Comment (mynktl) [reply src/istgt_iscsi.c]: ok

File: src/istgt_lu.c (hunk 5 above)
Comment (mynktl) [reply src/istgt_lu.c]: I have added this due to compilation warning. It seems we can remove this now since we are assigning value `0` to `lun`.

File: src/istgt_lu.c (hunk 7 above)
Comment (mynktl) [reply src/istgt_lu.c]: ok

File: src/istgt_lu_ctl.c (hunk 8 above)
Comment (mynktl) [reply src/istgt_lu_ctl.c]: ok.

File: src/istgt_lu_ctl.c (hunk 9 above)
Comment (mynktl) [reply src/istgt_lu_ctl.c]: ok

File: src/istgt_lu_disk.c (hunk 14 above)
Comment (mynktl) [reply src/istgt_lu_disk.c:119]: Ok, I will change it accordingly.

File: src/istgt_lu_disk.c (hunk 10 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/istgt_lu_disk.c (hunk 11 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: Ok. But if one LUN config has an error then we can still start ISTGT, right?
By returning an error here, we are not allowing ISTGT to start at all.
And `istgt_get_disktype_by_ext` will return `RAW` if `spec->file` has no file extension, so this `else` cond. can be removed.

File: src/istgt_lu_disk.c (hunk 12 above)
Comment (mynktl) [reply src/istgt_lu_disk.c:4177]: I removed this since `istgt_lu_disk_copy_reservation` always returns `0`.

File: src/istgt_lu_disk.c (hunk 13 above)
Comment (mynktl) [reply src/istgt_lu_disk.c:5126]: ok

File: src/istgt_misc.c (hunk 18 above)
Comment (mynktl) [reply src/istgt_misc.c]: ok

File: src/replication.c (hunk 19 above)
Comment (mynktl) [reply src/replication.c:819]: ok

File: src/replication.c (hunk 20 above)
Comment (mynktl) [reply src/replication.c:972]: yes. In `destroy_mempool`, we are handling this case if `cmdq` is not initialized.
To make it generic, I have kept it here.

File: src/replication.c (hunk 21 above)
Comment (mynktl) [reply src/replication.c]: ok

File: src/istgt.c [hunk 22]
```diff
@@ -2927,6 +2938,11 @@ main(int argc, char **argv)
 	if (rc < 0) {
 		ISTGT_ERRLOG("istgt_init() failed\n");
 	initialize_error:
+#ifdef	REPLICATION
```
Comment (vishnuitta) [inline src/istgt.c]: Lets avoid fixing these once-in-a-while leaks for the sake of stability. Also, reviewing this in holistic way takes lot of time.

File: src/istgt_iscsi.c [hunk 23]
```diff
@@ -7032,12 +7055,16 @@ istgt_iscsi_drop_all_conns(CONN_Ptr conn)
 			num++;
 		}
 	}
-	if(num!=0){
+
+	MTX_UNLOCK(&g_conns_mutex);
```
Comment (vishnuitta) [inline src/istgt_iscsi.c]: 'unlock'ing and 'lock' can be removed. (in next function also)

File: src/istgt_lu_disk.c [hunk 24]
```diff
@@ -952,15 +954,15 @@ istgt_lu_disk_init(ISTGT_Ptr istgt __attribute__((__unused__)), ISTGT_LU_Ptr lu)
 		rc = pthread_cond_init(&spec->cmd_queue_cond, NULL);
 		if (rc != 0) {
 			ISTGT_ERRLOG("LU%d: cond_init() failed errno:%d\n", lu->num, errno);
-			return -1;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: lets avoid error handling in this function, and, just return -1

File: src/istgt_lu_disk.c (hunk 13 above)
Comment (vishnuitta) [reply src/istgt_lu_disk.c:5126]: lets initialize initiator_ports to NULL when it got assigned to other structure variable

File: src/istgt_lu_disk.c [hunk 25]
```diff
@@ -8680,6 +8719,8 @@ istgt_lu_disk_queue_start(ISTGT_LU_Ptr lu, int lun, int worker_id)
 		MTX_UNLOCK(&conn->result_queue_mutex);
 	}
 
+	msg = NULL;
+	retval = 0;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: retval should be set to -1

File: src/ring_mempool.c [hunk 26]
```diff
@@ -76,6 +76,9 @@ destroy_mempool(rte_smempool_t *obj)
 	void *entry;
 	int rc;
 
+	if (!obj->ring)
```
Comment (vishnuitta) [inline src/ring_mempool.c]: lets have NULL check for obj

File: src/istgt.c (hunk 22 above)
Comment (mynktl) [reply src/istgt.c]: ok

File: src/istgt_iscsi.c (hunk 23 above)
Comment (mynktl) [reply src/istgt_iscsi.c]: ok

File: src/istgt_lu.c (hunk 6 above)
Comment (mynktl) [reply src/istgt_lu.c]: ok. will remove this.

File: src/istgt_lu_disk.c (hunk 24 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/istgt_lu_disk.c (hunk 13 above)
Comment (mynktl) [reply src/istgt_lu_disk.c:5126]: ok

File: src/istgt_lu_disk.c (hunk 15 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/istgt_lu_disk.c (hunk 16 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/istgt_lu_disk.c (hunk 25 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/istgt_lu_disk.c (hunk 17 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: ok

File: src/ring_mempool.c (hunk 26 above)
Comment (mynktl) [reply src/ring_mempool.c]: ok

**Comment by mynktl** [comment]
@vishnuitta  I have updated the changes according to review comments.
I have also added some required changes from PR #55  according to review comments.
Please have a look.
Thanks.

File: src/istgt_lu_disk.c [hunk 27]
```diff
@@ -5736,7 +5751,9 @@ istgt_lu_disk_lbwrite(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr lu_cm
 	if (markedForReturn == 1) {
 		ISTGT_ERRLOG("c#%d Error in entering block call", conn->id);
 		goto freeiovcnt;
-	}
+	} else if (markedForReturn == 2)
+		return 0;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: goto freeiovcnt is required.. add logs for markedForReturn == 2 case

File: src/istgt_lu_disk.c [hunk 28]
```diff
@@ -8538,6 +8568,7 @@ istgt_lu_disk_cmd_queue_count(ISTGT_LU_Ptr lu, int *lun)
 				}\
 			}\
 			MTX_UNLOCK(&spec->lu_tmf_mutex[worker_id]);\
+			decrement_conn_inflight = 0;		\
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: why this change?

Comment (mynktl) [reply src/istgt_lu_disk.c]: ok. we can remove this.

File: src/istgt_lu_disk.c (hunk 27 above)
Comment (mynktl) [reply src/istgt_lu_disk.c]: Ok.

File: src/istgt_lu_disk.c [hunk 29]
```diff
@@ -5406,8 +5407,10 @@ istgt_lu_disk_unmap(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr lu_cmd,
 			if (markedForReturn == 1) {
 				ret = -1;
 				break;
-			} else if (markedForReturn == 2)
+			} else if (markedForReturn == 2) {
+				errno = EBUSY;
 				return 0;
```
Comment (vishnuitta) [inline src/istgt_lu_disk.c]: we should be returning -1 from here..?

**Review by vishnuitta — APPROVED** [review]
changes are good. The last comment is related to unmap, which cStor currently doesn't support.

**Review by vishnuitta — APPROVED** [review]
Good..
